### ISSUE TITLE
deal with old unused FDs in lrucache

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#4427]
https://github.com/borgbackup/borg/issues/4360#issuecomment-467780888

as noted there, there might be a problem with old, open, but idle / unused FDs in the lrucache.

if we try to fetch an FD for a specific segment from the lrucache, we check if it's old and if so, we discard it (and the discard code also closes it) and rather reopen the segment file to avoid issues with stale file handles.

but if an FD just sits in the the cache for a long time, it might unnecessarily bind resources (not much within borg, but could be much in the filesystem's cache as seen in the davfs discussion).

ideas:
- borg usually works rather linearly, always appending stuff at the end when creating an archive and reading over it linearly when extracting an archive.
- exception: borg's fuse fs (borg mount): archive metadata is read once in a linear way, but we can't predict how content data is read
- also repo.list is not in on-disk order, repo.scan is.
- having a lot of old and idle FDs could be reduced by reducing the lrucache size, but considering fuse mounts and other non-linear access that is maybe not a good idea.
- we could also scan all lrucache entries and discard old stuff every time we fetch an entry from there if N seconds have elapsed since last fetch.

**Comment by fantasya-pbem** [comment]
I'll be happy to test changes that may improve DavFS cache usage!

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem see #4432. with the default of MAX_FD_AGE of 240s, no FD should age older than ~270s.

**Comment by fantasya-pbem** [comment]
I tested borgbackup version 1.2.0a4 with my 31 GB repo from [#4360]( https://github.com/borgbackup/borg/issues/4360#issuecomment-467757006). TLDR: No difference.

DavFS cache was 5 GB, but went up to 30 GB as described before. The whole check took ~ 51 min. No problems.

I noted that the message "Failed to securely erase old repository config file..." is new, didn't see that before alpha-4:

```
Found 93 segments
Failed to securely erase old repository config file (hardlinks not supported>). Old repokey data, if any, might persist on physical storage.
Checking segments
finished segment check at segment 92
Failed to securely erase old repository config file (hardlinks not supported>). Old repokey data, if any, might persist on physical storage.
Starting repository index check
Finished full repository check, no problems found.
```

**Comment by infectormp** [comment]
According to change log this fix was not included to 1.2.0a4

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem i pointed you to #4432, which is a pull request that is not merged yet into master (nor did it get into a release yet). can you fetch the related branch to test that code?

**Comment by fantasya-pbem** [comment]
Sorry, didn't check that, because alpha-4 was released too and I just used that without thinking... Will test the related branch soon.

**Comment by fantasya-pbem** [comment]
I tested your change. It seems to work as intended. DavFS cache didn't exceed 5.0 GB this time. The check took 45 minutes to complete without errors.

**Comment by fantasya-pbem** [comment]
Shall I test again with my old production repo ~ 117 GB? As mentioned in #4360 borg check crashed with DavFS cache 5 GB but succeeded with a bigger cache.

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem yeah, would be useful. but use a copy of the repo, don't run it against production repos.

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem while you're at it ... could you also test borg create and extract (with a reasonable amount of data so borg is busy for at least half an hour or so)?

**Comment by fantasya-pbem** [comment]
I tested borg check with my 117 GB repo too. No problems. Check took almost 5½ hours with a 5 GB DavFS cache.

After that I created a new test repo ([#4360, first comment](like https://github.com/borgbackup/borg/issues/4360#issuecomment-467506988)). This took 1h 43min without problems. A borg check took 40 minutes.

From my DavFS point of view, this change works and is a great improvement. I now have the choice of using small or big DavFS caches.
